packages/core/lib/config-migration.js

```diff
@@ -0,0 +1,51 @@
+const OS = require("os");
+const path = require("path");
+const fse = require("fs-extra");
+const Conf = require("conf");
+const { promisify } = require("util");
+const copy = require("./copy");
+const TruffleConfig = require("@truffle/config");
+
+module.exports = {
+  oldTruffleDataDirectory: path.join(OS.homedir(), ".config", "truffle"),
+
+  needsMigrated: function () {
+    const oldConfig = path.join(this.oldTruffleDataDirectory, "config.json");
+    const conf = new Conf({ projectName: "truffle" });
+    const alreadyMigrated = conf.get("migrated");
+    if (alreadyMigrated) return false;
+    return fse.existsSync(oldConfig) && oldConfig !== conf.path;
+  },
+
+  migrateTruffleDataIfNecessary: async function () {
+    if (!this.needsMigrated()) return;
+    this.migrateGlobalConfig();
+    const folders = ["compilers", ".db"];
+    for (const folder of folders) {
+      await this.migrateFolder(folder);
+    }
+  },
+
+  migrateGlobalConfig: function () {
+    const conf = new Conf({ projectName: "truffle" });
+    // set this to true so we know not to migrate next time
+    conf.set("migrated", true);
```

Comment:
I'm worried that `migrated` can be set and subsequent migration can fail which would leave "config" in an unknown state. Can `migrated` be set iff the migration is successful?

Reply:
hm, so thinking about this more - `migrated` isn't necessary, even, right? cause if it's in the new location, it's migrated; if it's in the old location, it's not. do we need more than that?

Reply:
Ok, so this is true! I thought that the new config was created on instantiation of the new config (`new Conf({ ...`) but it actually doesn't create the file until a field is set. So I think the logic for `needsMigrated` can be simplified too.

Reply:
I think "migration finished" is synonymous with "the file in the new location is a valid JSON file"

Reply:
@gnidan the operation isn't currently atomic, so there might be an issue if migration doesn't run to successful completion.

Reply:
@eggplantzzz Perhaps **migrate** is confirmation the migration finished successfully.

Reply:
> @gnidan the operation isn't currently atomic, so there might be an issue if migration doesn't run to successful completion.

Sure, but even if it's not atomic, what we care about is whether it's completed or not - if the file is valid and in a new location, that's enough information to say that migration is complete? Or am I missing something?

Reply:
My concern is if [this logic fails](https://github.com/trufflesuite/truffle/blob/4930591364f1126a9f9773178e2111d981a76024/packages/core/lib/config-migration.js#L24-L26) . `.db` and or `compilers` could be incomplete, and I'm not sure what a partially copied `.db` means.

Reply:
Yeah so we should figure out what the criteria for this is. And is it guaranteed that you'll even have `config.json`? You may just have `/compilers`. Maybe this is overkill and we should just check if the directory exists (the new directory) and if it is assume everything is migrated.